File: bootstrap/src/main/java/ru/hollowhorizon/hollowengine/bootstrap/impl/EmbeddedRuntimeJar.java
package ru.hollowhorizon.hollowengine.bootstrap.impl;

import org.jetbrains.annotations.Nullable;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Comparator;

final class EmbeddedRuntimeJar {
    private static final String RUNTIME_JAR = "META-INF/hollowengine/runtime/HollowEngineRuntime.jar";
    private static final String RUNTIME_SHA = "META-INF/hollowengine/runtime/HollowEngineRuntime.sha256";

    private EmbeddedRuntimeJar() {
    }

    static @Nullable File extract(Class<?> anchor, File cacheDirectory) throws IOException {
        ClassLoader classLoader = anchor.getClassLoader();
        String checksum;
        try (InputStream shaStream = classLoader.getResourceAsStream(RUNTIME_SHA)) {
            if (shaStream == null) return null;

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(shaStream, StandardCharsets.UTF_8))) {
                checksum = reader.readLine();
            }
        }

        if (checksum == null || checksum.isBlank()) return null;

        Path runtimeDir = cacheDirectory.toPath().resolve("runtime");
        Files.createDirectories(runtimeDir);
        Path target = runtimeDir.resolve("HollowEngineRuntime-" + checksum + ".jar");
        if (Files.exists(target)) return target.toFile();

        cleanDirectory(cacheDirectory);

        try (InputStream jarStream = classLoader.getResourceAsStream(RUNTIME_JAR)) {
            if (jarStream == null) return null;
            Files.createDirectory(target.getParent());
            Files.createFile(target);
            Files.copy(jarStream, target, StandardCopyOption.REPLACE_EXISTING);
        }

        return target.toFile();
    }

    private static void cleanDirectory(File file) throws IOException {
        var path = file.toPath();
        try (var stream = Files.walk(path)) {
            stream.sorted(Comparator.reverseOrder())
                    .filter(p -> !p.equals(path))
                    .forEach(p -> {
                        try {
                            Files.delete(p);
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    });
        }
    }
}

File: runtime/src/main/java/ru/hollowhorizon/hollowengine/common/runtime/EmbeddedRuntimeJar.kt
package ru.hollowhorizon.hollowengine.common.runtime

import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import kotlin.io.path.createDirectories
import kotlin.io.path.exists

object EmbeddedRuntimeJar {
    private const val RUNTIME_JAR_RESOURCE = "META-INF/hollowengine/runtime/HollowEngineRuntime.jar"
    private const val RUNTIME_SHA_RESOURCE = "META-INF/hollowengine/runtime/HollowEngineRuntime.sha256"

    fun extract(anchor: Class<*>, cacheDirectory: File): File? {
        val classLoader = anchor.classLoader
        val checksum = classLoader.getResourceAsStream(RUNTIME_SHA_RESOURCE)
            ?.bufferedReader()
            ?.use { it.readText().trim() }
            ?.takeIf(String::isNotBlank)
            ?: return null

        val runtimeDirectory = cacheDirectory.toPath().resolve("runtime").apply { createDirectories() }
        val target = runtimeDirectory.resolve("HollowEngineRuntime-$checksum.jar")
        if (target.exists()) return target.toFile()

        val temp = runtimeDirectory.resolve("HollowEngineRuntime-$checksum.jar.tmp")
        classLoader.getResourceAsStream(RUNTIME_JAR_RESOURCE)?.use { input ->
            Files.copy(input, temp, StandardCopyOption.REPLACE_EXISTING)
        } ?: return null

        Files.move(temp, target, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE)
        return target.toFile()
    }
}
